examples/pytorch/graphsage_base/node_classification.py

```diff
@@ -0,0 +1,141 @@
+            feat = feat.to(device)
+            for input_nodes, output_nodes, blocks in tqdm.tqdm(dataloader):
+                # use an explicitly contiguous slice
+                x = feat[input_nodes]
+                h = layer(blocks[0], x)
+                if l != len(self.layers) - 1:
+                    h = F.relu(h)
+                    h = self.dropout(h)
+                # by design, our output nodes are contiguous so we can take
+                # advantage of that here
+                y[output_nodes[0]:output_nodes[-1]+1] = h.to(buffer_device)
+            feat = y
+        return y
+    
+def evaluate(model, graph, dataloader):
+    model.eval()
+    ys = []
+    y_hats = []
+    for it, (input_nodes, output_nodes, blocks) in enumerate(dataloader):
+        with torch.no_grad():
+            x = blocks[0].srcdata['feat']
+            ys.append(blocks[-1].dstdata['label'])
+            y_hats.append(model(blocks, x))
+    return MF.accuracy(torch.cat(y_hats), torch.cat(ys))
+
+def layerwise_infer(args, device, graph, nid, model, batch_size):
+    model.eval()
+    nid = nid.to(device)
+    with torch.no_grad():
+        pred = model.inference(graph, device, batch_size)
+        pred = pred[nid].to(device)
+        label = graph.ndata['label'][nid].to(device)
+        return MF.accuracy(pred, label)
+
+def train(args, device, g, dataset, model):
+    # create sampler & dataloader
+    train_idx = dataset.train_idx.to(device)
+    val_idx = dataset.val_idx.to(device)
+    sampler = dgl.dataloading.NeighborSampler([15, 10, 5],
+                                              prefetch_node_feats=['feat'],
+                                              prefetch_labels=['label'])
+    use_uva = (args.mode == 'mixed')
+    train_dataloader = dgl.dataloading.DataLoader(g, train_idx, sampler, device=device,
+                                                  batch_size=1024, shuffle=True,
+                                                  drop_last=False, num_workers=0,
+                                                  use_uva=use_uva)
+
+    valid_dataloader = dgl.dataloading.DataLoader(g, val_idx, sampler, device=device,
+                                                  batch_size=1024, shuffle=True,
+                                                  drop_last=False, num_workers=0,
+                                                  use_uva=use_uva)
+
+    opt = torch.optim.Adam(model.parameters(), lr=1e-3, weight_decay=5e-4)
+    
+    for epoch in range(10):
+        model.train()
+        total_loss = torch.tensor(0.0).to(device)
+        for it, (input_nodes, output_nodes, blocks) in enumerate(train_dataloader):
+            x = blocks[0].srcdata['feat']
+            y = blocks[-1].dstdata['label']
+            y_hat = model(blocks, x)
+            loss = F.cross_entropy(y_hat, y)
+            opt.zero_grad()
+            loss.backward()
+            opt.step()
+            total_loss += loss

```

Comment:
I remember using `loss` rather than `loss.item()` caused memory leak issues. I'm not sure if this is still the case.